train.py

```diff
@@ -0,0 +1,491 @@
+                        help='Root directory of dataset')
+    parser.add_argument('--identifier', type=str, default='mapillary_nopanos',
+                        help='Description of this model, e.g. mapillary_nopanos_vgg16_netvlad')
+    parser.add_argument('--nEpochs', type=int, default=30, help='number of epochs to train for')
+    parser.add_argument('--start_epoch', default=0, type=int, metavar='N',
+                        help='manual epoch number (useful on restarts)')
+    parser.add_argument('--save_every_epoch', action='store_true', help='Flag to set a separate checkpoint file for each new epoch')
+    parser.add_argument('--threads', type=int, default=6, help='Number of threads for each data loader to use')
+    parser.add_argument('--nocuda', action='store_true', help='If true, use CPU only. Else use GPU.')
+
+
+    opt = parser.parse_args()
+    print(opt)
+
+    # TODO: add some kind of check if a resumed ckpt has different parameters compared to the train.ini file
+    # TODO: implement training on pitts
+
+    configfile = opt.config_path
+    assert os.path.isfile(configfile)
+    config = configparser.ConfigParser()
+    config.read(configfile)
+
+    cuda = not opt.nocuda
+    if cuda and not torch.cuda.is_available():
+        raise Exception("No GPU found, please run with --nocuda")
+
+    device = torch.device("cuda" if cuda else "cpu")
+
+    random.seed(int(config['train']['seed']))
+    np.random.seed(int(config['train']['seed']))
+    torch.manual_seed(int(config['train']['seed']))
+    if cuda:
+        # noinspection PyUnresolvedReferences
+        torch.cuda.manual_seed(int(config['train']['seed']))
+
+    optimizer = None
+    scheduler = None
+
+    print('===> Building model')
+
+    encoder_dim, encoder = get_backend()
+
+    if opt.resume_path: # if already started training earlier and continuing
+        if isfile(opt.resume_path):
+            print("=> loading checkpoint '{}'".format(opt.resume_path))
+            checkpoint = torch.load(opt.resume_path, map_location=lambda storage, loc: storage)
+            config['global_params']['num_clusters'] = str(checkpoint['state_dict']['pool.centroids'].shape[0])
+
+            model = get_model(encoder, encoder_dim, opt, config['global_params'], append_pca_layer=False)
+
+            model.load_state_dict(checkpoint['state_dict'])
+            opt.start_epoch = checkpoint['epoch']
+
+            print("=> loaded checkpoint '{}'".format(opt.resume_path, ))
+        else:
+            raise FileNotFoundError("=> no checkpoint found at '{}'".format(opt.resume_path))
+    else: # if not, assume fresh training instance and will initially generate cluster centroids
+        print('===> Finding cluster centroids')
+
+        print('===> Loading dataset(s) for clustering')
+        train_dataset = MSLS(opt.dataset_root_dir, mode='test', cities='train', transform=input_transform(),
```

Comment:
I think this is only needed in the `else` part of the `if opt.cluster_path:` statement?

Reply:
Good point, I moved it